### PR TITLE
Redesign of the association maps, multivector manager, HGCAL Rechits and Validation with significant speedup of Phase-2 workflows

### PR DESCRIPTION
**Pull request by felicepantaleo**
#### PR description:

In the context of Next Generation Triggers and HGCAL TICL reconstruction, fast AssociatorMaps between Reco and Sim, Sim to Sim, and Reco to Reco are required in order to develop new reconstruction algorithms and study their performance.  They have also been included in the HGCalValidator and in the TICLDumper, but can be extended to many other detectors (MTD for instance, as it is the new main offender). 
The HGCAL Rechit producer was optimized by avoiding calling virtual functions for every rechit and to enable support for the MultiVectorManager.
The MultiVectorManager was optimized and together with the HGCAL RecHit map was used to optimize the E/Gamma reconstruction and the associators.
The HGCAL Validation was redesigned to make effective usage of the new associators. The framework was completely redesigned to generate automatically validation plots for any new collection of TICL tracksters  against SimTracksters from Simclusters and CaloParticles. 

#### Performance measurements:
Performed with `14_1_0_pre7` with TTbar PU200, 10 events, single thread, single stream

[**14_1_0_pre7**](https://felice.web.cern.ch/CMS/circles/web/piechart.php?local=false&dataset=Phase2Timing_resources_14_1_0_pre7&resource=time_thread&colours=default&groups=consolidated_group_mapping&show_labels=true&threshold=0): 

- Total time: 536s
- Total allocated memory: 90GB

[**14_1_0_pre7+PR**](https://felice.web.cern.ch/CMS/circles/web/piechart.php?local=false&dataset=Phase2Timing_resources_newAssociators&resource=time_thread&colours=default&groups=consolidated_group_mapping&show_labels=true&threshold=0)

- Total time: 171s
- Total allocated memory: 56GB

#### Performance comparison 

Full performance comparison table here: 
https://docs.google.com/spreadsheets/d/1vdmDDJJ07tf9ekQ8EC24sMqqvtCtiU2Mh6FawsIRu24/edit?usp=sharing

Notable numbers:
- Overall Phase-2 TTbar PU200 Reconstruction+Validation workflows: speedup **313%** , allocated memory **-38%**
- Overall E/Gamma Phase-2 Reconstruction: **1.5x** speedup
- ElectronSeedProducer ecalDrivenElectronSeeds: **3x** speedup
- HGCalValidator: **30x** speedup
- New associators: between **two orders of magnitude** and infinite speedup depending if you want to run on the CaloParticles from pileup as well. In this case the associators in release will crash for the amount of resources needed.

- Job Memory usage step4:  -22%
image

@cms-sw/hgcal-dpg-l2 @cms-sw/egamma-pog-l2 @waredjeb @AuroraPerego @rovere

### CODE REVIEW

**Comment by felicepantaleo** [comment]
test parameters:  

- workflow_opts= -w upgrade
- workflow = 29896.203

**Comment by felicepantaleo** [comment]
@cmsbuild please test

**Comment by felicepantaleo** [comment]
You can find the full details of the performance comparison here:
https://docs.google.com/spreadsheets/d/1vdmDDJJ07tf9ekQ8EC24sMqqvtCtiU2Mh6FawsIRu24/edit?usp=sharing

I'll add this link in the initial message as well.

**Comment by felicepantaleo** [comment]
@cmsbuild please test

**Comment by felicepantaleo** [comment]
@cmsbuild please test

**Comment by felicepantaleo** [comment]
@cmsbuild please test

**Comment by felicepantaleo** [comment]
enable profiling

**Comment by felicepantaleo** [comment]
I'm enabling profiling even though the validation is not run in profiling workflows. Is it possible to specify another workflow for profiling?

**Comment by felicepantaleo** [comment]
@cmsbuild please test

**Comment by jfernan2** [comment]
Profiling comparison tests run on wfs  29834.21 (D110 upgrade) and 12634.21 (Run3 2023) as set on https://github.com/cms-sw/cms-bot/pull/2282

However, the timing comparison fails due to an unknown igprof (and VTune too) segmentation fault, this is a long standing problem which prevents to have timing studies in PRs unfortunately

https://github.com/cms-sw/cmssw/issues/43166

**Comment by felicepantaleo** [comment]
Thanks @jfernan2 for clarifying. Unfortunately 29834.21 does not run any validation nor prevalidation during step3.

File: RecoHGCal/TICL/plugins/TICLDumper.cc [hunk 1]
```diff
@@ -1442,4 +1415,4 @@ void TICLDumper::fillDescriptions(edm::ConfigurationDescriptions& descriptions)
   descriptions.add("ticlDumper", desc);
 }
 
-DEFINE_FWK_MODULE(TICLDumper);
+DEFINE_FWK_MODULE(TICLDumper);
```
Comment (civanch) [inline RecoHGCal/TICL/plugins/TICLDumper.cc:1418]: Here there is a hidden symbol.

File: RecoLocalCalo/HGCalRecAlgos/src/ClusterTools.cc [hunk 2]
```diff
@@ -132,21 +122,27 @@ bool ClusterTools::getWidths(const reco::CaloCluster& clus,
...
+      float hitEnergy = rhmanager[rechitIndex].energy();
 
       GlobalPoint cellPos = rhtools_.getPosition(id);
-      double weight = theHit->energy();
+      double weight = hitEnergy;
       // take w0=2 To be optimized
       double logweight = 0;
       if (clus.energy() != 0) {
-        logweight = std::max(0., 2 + log(theHit->energy() / clus.energy()));
+        logweight = std::max(0., 2 + log(hitEnergy / clus.energy()));
```
Comment (civanch) [inline RecoLocalCalo/HGCalRecAlgos/src/ClusterTools.cc:145]: should be std::log ?

File: RecoLocalCalo/HGCalRecProducers/plugins/HGCalRecHitWorkerSimple.cc [hunk 3]
```diff
@@ -121,114 +121,114 @@ void HGCalRecHitWorkerSimple::set(const edm::EventSetup& es) {
...
+
+    if (detid.det() == DetId::Forward || detid.det() == DetId::Hcal) {
+      if ((detid & rangeMask_) == rangeMatch_)
+        continue;
+    }
+
+    int thickness = -1;
+    float sigmaNoiseGeV = 0.f;
+    unsigned int layer = tools_->getLayerWithOffset(detid);
+    float cce_correction = 1.0;
```
Comment (civanch) [inline RecoLocalCalo/HGCalRecProducers/plugins/HGCalRecHitWorkerSimple.cc:139]: should be 1.f ?

File: SimCalorimetry/HGCalAssociatorProducers/plugins/HitToTracksterAssociatorProducer.h [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+private:
+  void produce(edm::StreamID, edm::Event &, const edm::EventSetup &) const override;
+
+  edm::EDGetTokenT<std::vector<reco::CaloCluster>> LCCollectionToken_;
+  edm::EDGetTokenT<std::vector<ticl::Trackster>> tracksterCollectionToken_;
+  edm::EDGetTokenT<std::unordered_map<DetId, const unsigned int>> hitMapToken_;
+  std::vector<edm::EDGetTokenT<HGCRecHitCollection>> hitsTokens_;
+};
+
+#endif
```
Comment (civanch) [inline SimCalorimetry/HGCalAssociatorProducers/plugins/HitToTracksterAssociatorProducer.h:37]: Hidden symbol

**Comment by civanch** [comment]
@felicepantaleo , rebase is required.

**Comment by felicepantaleo** [comment]
thanks @civanch. I have addressed all your comments.
I have also optimized the Pattern Recognition by Recovery.

**Comment by felicepantaleo** [comment]
please test

**Comment by civanch** [comment]
+simulation

**Comment by jfernan2** [comment]
enable profiling
Launching tests once more to get last FastTimer CPU profiling changes from:
https://github.com/cms-sw/cms-bot/pull/2348
https://github.com/cms-sw/cms-bot/pull/2349
https://github.com/cms-sw/cms-bot/pull/2343
https://github.com/cms-sw/cms-bot/pull/2344

**Comment by jfernan2** [comment]
please test

**Comment by jfernan2** [comment]
@felicepantaleo we finally have a proper profiing comparison:

https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-232398/41851/profiling/29834.21/diff-step3_cpu.resources.json.html

I see HGCalRecHitProducer reducing a 85% the timing wow!! I guess this is compatible with the timing reduction you saw in your tests, thanks!

**Comment by jfernan2** [comment]
+1

**Comment by Moanwar** [comment]
+Upgrade

Indeed, thanks for these good results and the updates.

**Comment by mmusich** [comment]
enable hlt_p2_timing

**Comment by mmusich** [comment]
@cmsbuild, please test

**Comment by mmusich** [comment]
+hlt
   * see https://github.com/cms-sw/cmssw/pull/45865#issuecomment-2388364253
   * see https://github.com/cms-sw/cmssw/pull/45865#issuecomment-2388717635

**Comment by felicepantaleo** [comment]
> book-keeping I'd like to test it once again enabling the hlt phase-2 timing flag once we have an IB that contains #46185.

@mmusich thanks. Do you think it is worth **also** nesting `hlt_p2_timing` into `profiling`, so that when running with profiling enabled we also get the hlt_p2_timing enabled?

**Comment by mmusich** [comment]
> Do you think it is worth also nesting hlt_p2_timing into profiling, so that when running with profiling enabled we also get the hlt_p2_timing enabled?

no strong opinion, if that's fine with @cms-sw/reconstruction-l2 it's fine by me. OTOH I would reserve the option of running *just* the phase-2 HLT menu timing without the rest of the reco profiling.

**Comment by jfernan2** [comment]
H

> > Do you think it is worth also nesting hlt_p2_timing into profiling, so that when running with profiling enabled we also get the hlt_p2_timing enabled?
> 
> no strong opinion, if that's fine with @cms-sw/reconstruction-l2 it's fine by me. OTOH I would reserve the option of running _just_ the phase-2 HLT menu timing without the rest of the reco profiling.

Honestly, I'd rather prefer to keep the two tests decoupled for the moment, we are refurbishing the profiling test at present due to igprof problem. Thanks

**Comment by felicepantaleo** [comment]
@cms-sw/dqm-l2 let me know if you have any more comments.

**Comment by mmusich** [comment]
pinging again @cms-sw/dqm-l2 @antoniovagnerini, please review / sign, thank you.

File: Validation/HGCalValidation/interface/HGVHistoProducerAlgo.h [hunk 5]
```diff
@@ -142,57 +143,58 @@ struct HGVHistoProducerAlgoHistograms {
...
+  std::vector<dqm::reco::MonitorElement*> h_num_caloparticle_en[numberOfValidationTypes_];
+  std::vector<dqm::reco::MonitorElement*> h_num_caloparticle_pt[numberOfValidationTypes_];
+  std::vector<dqm::reco::MonitorElement*> h_numDup_trackster_eta[numberOfValidationTypes_];
+  std::vector<dqm::reco::MonitorElement*> h_numDup_trackster_phi[numberOfValidationTypes_];
+  std::vector<dqm::reco::MonitorElement*> h_numDup_trackster_en[numberOfValidationTypes_];
+  std::vector<dqm::reco::MonitorElement*> h_numDup_trackster_pt[numberOfValidationTypes_];
+  std::vector<dqm::reco::MonitorElement*> h_denom_caloparticle_eta[numberOfValidationTypes_];
+  std::vector<dqm::reco::MonitorElement*> h_denom_caloparticle_phi[numberOfValidationTypes_];
+  std::vector<dqm::reco::MonitorElement*> h_denom_caloparticle_en[numberOfValidationTypes_];
+  std::vector<dqm::reco::MonitorElement*> h_denom_caloparticle_pt[numberOfValidationTypes_];
```
Comment (antoniovagnerini) [inline Validation/HGCalValidation/interface/HGVHistoProducerAlgo.h:197]: can you confirm that the doubling in `numberOfValidationTypes_` from 2 to 4 is behind the expected increase in DQM Histogram size by 2.7MB? 
`DQMHistoSizes: Histogram memory added: 2745.032 KiB( 44 files compared)`

Comment (felicepantaleo) [reply Validation/HGCalValidation/interface/HGVHistoProducerAlgo.h:197]: yep

**Comment by antoniovagnerini** [comment]
+1

**Comment by mandrenguyen** [comment]
+1

**Pull request merged**